drivers/ethernet/eth_stm32_hal_v2.c

```diff
@@ -0,0 +1,637 @@
+/*
+ * Copyright (c) 2017 Erwin Rol <erwin@erwinrol.com>
+ * Copyright (c) 2020 Alexander Kozhinov <ak.alexander.kozhinov@gmail.com>
+ * Copyright (c) 2021 Carbon Robotics
+ * Copyright (c) 2025 STMicroelectronics
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/util.h>
+#include <errno.h>
+#include <stdbool.h>
+#include <zephyr/net/ethernet.h>
+#include <zephyr/net/phy.h>
+#include <ethernet/eth_stats.h>
+#include <soc.h>
+#include <zephyr/irq.h>
+
+#if defined(CONFIG_PTP_CLOCK_STM32_HAL)
+#include <zephyr/drivers/ptp_clock.h>
+#endif /* CONFIG_PTP_CLOCK_STM32_HAL */
+
+#include "eth.h"
+#include "eth_stm32_hal_priv.h"
+
+#define LOG_MODULE_NAME eth_stm32_hal
+
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32n6_ethernet)
+#define ETH_GMII_MODE	HAL_ETH_GMII_MODE
+#define ETH_RGMII_MODE	HAL_ETH_RGMII_MODE
+#endif
+
+#define ETH_DMA_TX_TIMEOUT_MS	20U  /* transmit timeout in milliseconds */
+
+struct eth_stm32_rx_buffer_header {
+	struct eth_stm32_rx_buffer_header *next;
+	uint16_t size;
+	bool used;
+};
+
+struct eth_stm32_tx_buffer_header {
+	ETH_BufferTypeDef tx_buff;
+	bool used;
+};
+
+static struct eth_stm32_rx_buffer_header dma_rx_buffer_header[ETH_RXBUFNB];
+static struct eth_stm32_tx_buffer_header dma_tx_buffer_header[ETH_TXBUFNB];
+static struct eth_stm32_tx_context dma_tx_context[ETH_TX_DESC_CNT];
+
+void HAL_ETH_RxAllocateCallback(uint8_t **buf)
+{
+	for (size_t i = 0; i < ETH_RXBUFNB; ++i) {
+		if (!dma_rx_buffer_header[i].used) {
+			dma_rx_buffer_header[i].next = NULL;
+			dma_rx_buffer_header[i].size = 0;
+			dma_rx_buffer_header[i].used = true;
+			*buf = dma_rx_buffer[i];
+			return;
+		}
+	}
+	*buf = NULL;
+}
+
+/* Pointer to an array of ETH_STM32_RX_BUF_SIZE uint8_t's */
+typedef uint8_t (*RxBufferPtr)[ETH_STM32_RX_BUF_SIZE];
```

Comment:
Could you move this definition before line 46?

Reply:
Done